src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs

```diff
@@ -0,0 +1,1620 @@
+                [InterceptsLocation("Program.cs", 15, 11)]
+                public static I1 Interceptor1(this I1 i1, string param) { Console.Write("interceptor " + param); return i1; }
+            }
+            """;
+        var verifier = CompileAndVerify(new[] { (source, "Program.cs"), s_attributesSource }, expectedOutput: "interceptor call siteinterceptable call site");
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void InterceptableFromMetadata()
+    {
+        var source1 = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            public class C
+            {
+                [Interceptable]
+                public C InterceptableMethod(string param) { Console.Write("interceptable " + param); return this; }
+            }
+            """;
+
+        var source2 = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            static class Program
+            {
+                public static void Main()
+                {
+                    var c = new C();
+                    c.InterceptableMethod("call site");
+                }
+            }
+
+            static class D
+            {
+                [InterceptsLocation("Program.cs", 9, 11)]
+                public static C Interceptor1(this C c, string param) { Console.Write("interceptor " + param); return c; }
+            }
+            """;
+
+        var comp1 = CreateCompilation(new[] { (source1, "File1.cs"), s_attributesSource });
+        comp1.VerifyEmitDiagnostics();
+
+        var verifier = CompileAndVerify((source2, "Program.cs"), references: new[] { comp1.ToMetadataReference() }, expectedOutput: "interceptor call site");
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void LocalFunctionInterceptor()
+    {
+        var source = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            interface I1 { }
+            class C : I1 { }
+
+            static class Program
+            {
+                [Interceptable]
+                public static void InterceptableMethod(string param) { Console.Write("interceptable " + param); }
+
+                public static void Main()
+                {
+                    InterceptableMethod("call site");
+
+                    [InterceptsLocation("Program.cs", 13, 8)]
+                    static void Interceptor1(string param) { Console.Write("interceptor " + param); }
+                }
+            }
+            """;
+        var comp = CreateCompilation(new[] { (source, "Program.cs"), s_attributesSource });
+        comp.VerifyDiagnostics(
+            // Program.cs(16,10): error CS27009: An interceptor method must be an ordinary member method.
+            //         [InterceptsLocation("Program.cs", 13, 8)]
+            Diagnostic(ErrorCode.ERR_InterceptorMethodMustBeOrdinary, @"InterceptsLocation(""Program.cs"", 13, 8)").WithLocation(16, 10),
```

Comment:
>ERR_InterceptorMethodMustBeOrdinary [](http://example.com/codeflow?start=33&length=35)

Consider testing the other kinds of functions (lambda, accessors, etc).

Reply:
Happy to expand on this a bit more, but not planning to test this exhaustively for now.